File: go/robot-name/robot_name.go
package robotname

import (
	"errors"
	"fmt"
)

type Robot struct {
	name  int
	named bool
}

const allNames = 26 * 26 * 1000

var names = map[int]bool{}
var namesAssigned = 0

const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

func (r *Robot) Name() (string, error) {
	if !r.named {
		if namesAssigned == allNames {
			return "", errors.New("all robot names have been assigned")
		}
		r.Reset()
	}
	return fmt.Sprintf("%c%c%03d", letters[r.name/26000], letters[(r.name%26000)/1000], r.name%1000), nil
}

func (r *Robot) Reset() {
	if len(names) == 0 {
		for i := 0; i < allNames; i++ {
			names[i] = false
		}
	}
	for name := range names {
		r.name = name
		break
	}
	delete(names, r.name)
	r.named = true
	namesAssigned++
}

File: clojure/robot-name/src/robot_name.clj
(ns robot-name)

(def letters ["A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X" "Y" "Z"])

(def used-names (atom #{}))

(defn generate-name []
  (let [name (str (rand-nth letters) (rand-nth letters) (rand-int 10) (rand-int 10) (rand-int 10))]
    (if (get @used-names name)
      (recur)
      (do
        (swap! used-names conj name)
        name))))

(defn robot []
  (atom {:name (generate-name)})
)

(defn robot-name [robot]
  (get @robot :name)
)

(defn reset-name [robot]
  (swap! robot #(assoc % :name (generate-name)))
)
